File: cogs/localai.py
import discord
from discord.ext import commands
from discord import app_commands
import aiohttp
import json
import random
from db.localai_db import LocalAIDB
from typing import Optional, List

class LocalAI(commands.Cog):
    chatbot_group = app_commands.Group(
        name="chatbot",
        description="Manage your local AI chatbot"
    )

    def __init__(self, bot):
        self.bot = bot
        self.db = LocalAIDB()
        self.session = aiohttp.ClientSession()
        self.mention_triggers = [f"<@{bot.user.id}>", f"<@!{bot.user.id}>"]

    def cog_unload(self):
        self.bot.loop.create_task(self.session.close())
        self.db.close()

    def is_mentioned(self, message: discord.Message) -> bool:
        """Check if bot is mentioned in message"""
        return any(trigger in message.content for trigger in self.mention_triggers)

    # Modify the get_ai_response method to include the system prompt
    async def get_ai_response(self, guild_id: int, prompt: str) -> Optional[str]:
        """Get response from local AI endpoint"""
        config = self.db.get_config(guild_id)
        if not config or not config['enabled']:
            return None
    
        headers = {}
        if config['api_key']:
            headers["Authorization"] = f"Bearer {config['api_key']}"
    
        # Get system prompt if configured
        system_prompt = self.db.get_system_prompt(guild_id)
        
        # Build messages array
        messages = []
        if system_prompt:
            messages.append({"role": "system", "content": system_prompt})
        messages.append({"role": "user", "content": prompt})
    
        payload = {
            "model": config['model_name'],
            "messages": messages,  # Use the messages array we built
            "temperature": config.get('temperature', 0.7),
            "max_tokens": config.get('max_tokens', 1000)
        }
    
        try:
            async with self.session.post(
                f"{config['api_base']}/v1/chat/completions",
                headers=headers,
                json=payload,
                timeout=aiohttp.ClientTimeout(total=30)
            ) as response:
                if response.status == 200:
                    data = await response.json()
                    return data['choices'][0]['message']['content']
                error = await response.text()
                print(f"AI API Error: {error}")
                return None
        except Exception as e:
            print(f"AI Request Failed: {str(e)}")
            return None

    @commands.Cog.listener()
    async def on_message(self, message: discord.Message):
        """Handle message events for AI responses"""
        if (
            message.author.bot or
            not message.guild or
            not isinstance(message.channel, discord.TextChannel)
        ):
            return

        # Always respond if bot is mentioned
        mentioned = self.is_mentioned(message)

        # Always respond if replying to the bot
        is_reply_to_bot = False
        if message.reference and isinstance(message.reference.resolved, discord.Message):
            replied_message = message.reference.resolved
            if replied_message.author.id == self.bot.user.id:
                is_reply_to_bot = True

        # Check whitelist
        whitelisted = self.db.is_whitelisted(message.guild.id, message.channel.id)

        # Decide if should respond
        should_respond = False
        if mentioned or is_reply_to_bot:
            should_respond = True
        elif whitelisted:
            chance = self.db.get_response_chance(message.guild.id)
            roll = random.uniform(0, 100)
            if roll <= chance:
                should_respond = True

        if not should_respond:
            return

        # Get clean prompt (remove mentions)
        prompt = message.clean_content
        for trigger in self.mention_triggers:
            prompt = prompt.replace(trigger, '').strip()

        if not prompt:
            return

        async with message.channel.typing():
            response = await self.get_ai_response(message.guild.id, prompt)
            if response:
                # Truncate if needed
                if len(response) > 2000:
                    response = response[:1997] + "..."
                await message.reply(response, mention_author=False)

    @chatbot_group.command(name="configure", description="Configure your local AI endpoint")
    @app_commands.describe(
        api_base="Your local API URL (e.g., http://localhost:1234)",
        api_key="API key if required (leave empty if none)",
        model_name="Model name in your LM Studio"
    )
    async def configure_local_ai(
        self,
        interaction: discord.Interaction,
        api_base: str,
        api_key: Optional[str] = None,
        model_name: str = "local-model"
    ):
        if not interaction.user.guild_permissions.administrator:
            await interaction.response.send_message("You do not have permission to use this command.", ephemeral=True)
            return

        if not api_base.startswith(('http://', 'https://')):
            return await interaction.response.send_message(
                "Invalid URL format. Must start with http:// or https://",
                ephemeral=True
            )

        self.db.set_config(interaction.guild.id, api_base, api_key, model_name)
        await interaction.response.send_message(
            f"Local AI configured!\n"
            f"Endpoint: `{api_base}`\n"
            f"Model: `{model_name}`\n"
            f"API Key: {'Configured' if api_key else 'None'}",
            ephemeral=True
        )

    @chatbot_group.command(name="prompt", description="Set a custom system prompt")
    @app_commands.describe(
        prompt="System prompt to prepend to all AI requests (leave empty to clear)"
    )
    async def set_system_prompt(
        self,
        interaction: discord.Interaction,
        prompt: Optional[str] = None
    ):
        if not interaction.user.guild_permissions.administrator:
            await interaction.response.send_message("You do not have permission to use this command.", ephemeral=True)
            return

        self.db.set_system_prompt(interaction.guild.id, prompt)
        if prompt:
            await interaction.response.send_message(
                f"System prompt set! It will be prepended to all AI requests.",
                ephemeral=True
            )
        else:
            await interaction.response.send_message(
                "System prompt cleared!",
                ephemeral=True
            )

    @chatbot_group.command(name="toggle", description="Enable or disable AI responses")
    @app_commands.describe(enabled="Enable/disable AI responses")
    async def toggle_ai(self, interaction: discord.Interaction, enabled: bool):
        if not interaction.user.guild_permissions.administrator:
            await interaction.response.send_message("You do not have permission to use this command.", ephemeral=True)
            return

        self.db.set_enabled(interaction.guild.id, enabled)
        await interaction.response.send_message(
            f"Local AI responses {'enabled' if enabled else 'disabled'}",
            ephemeral=True
        )

    @chatbot_group.command(name="channel", description="Add or remove a whitelisted channel")
    @app_commands.describe(
        channel="Channel to whitelist",
        action="Add or remove from whitelist"
    )
    async def manage_whitelist(
        self,
        interaction: discord.Interaction,
        channel: discord.TextChannel,
        action: str
    ):
        if not interaction.user.guild_permissions.administrator:
            await interaction.response.send_message("You do not have permission to use this command.", ephemeral=True)
            return

        if action.lower() == "add":
            self.db.add_whitelisted_channel(interaction.guild.id, channel.id)
            await interaction.response.send_message(
                f"Added {channel.mention} to whitelist",
                ephemeral=True
            )
        elif action.lower() == "remove":
            self.db.remove_whitelisted_channel(interaction.guild.id, channel.id)
            await interaction.response.send_message(
                f"Removed {channel.mention} from whitelist",
                ephemeral=True
            )
        else:
            await interaction.response.send_message(
                "Invalid action. Use 'add' or 'remove'",
                ephemeral=True
            )

    @chatbot_group.command(name="listchannels", description="List all whitelisted channels")
    async def list_whitelisted_channels(self, interaction: discord.Interaction):
        channels = self.db.get_whitelisted_channels(interaction.guild.id)
        if not channels:
            return await interaction.response.send_message(
                "No whitelisted channels",
                ephemeral=True
            )

        channel_mentions = []
        for channel_id in channels:
            channel = interaction.guild.get_channel(channel_id)
            if channel:
                channel_mentions.append(channel.mention)

        embed = discord.Embed(
            title="Whitelisted Channels",
            description="\n".join(channel_mentions) or "None",
            color=0x7289da
        )
        await interaction.response.send_message(embed=embed, ephemeral=True)

    @chatbot_group.command(name="chance", description="Set AI response chance percentage (0-100)")
    @app_commands.describe(
        chance="Chance percentage (0-100). 100 means always respond in whitelisted channels."
    )
    async def set_response_chance(
        self,
        interaction: discord.Interaction,
        chance: float
    ):
        if not interaction.user.guild_permissions.administrator:
            await interaction.response.send_message(
                "You do not have permission to use this command.",
                ephemeral=True
            )
            return

        if chance < 0 or chance > 100:
            await interaction.response.send_message(
                "Chance must be between 0 and 100.",
                ephemeral=True
            )
            return

        self.db.set_response_chance(interaction.guild.id, chance)
        await interaction.response.send_message(
            f"AI response chance set to {chance}%.",
            ephemeral=True
        )

async def setup(bot):
    await bot.add_cog(LocalAI(bot))

File: commands/localai.js
const { 
    SlashCommandBuilder, 
    PermissionFlagsBits,
    ChannelType,
    EmbedBuilder
} = require('discord.js');
const axios = require('axios');
const localAIModel = require('../models/localAIModel');

module.exports = {
    data: new SlashCommandBuilder()
        .setName('chatbot')
        .setDescription('Manage your local AI chatbot')
        .setDefaultMemberPermissions(PermissionFlagsBits.Administrator)
        .addSubcommand(subcommand =>
            subcommand
                .setName('configure')
                .setDescription('Configure your local AI endpoint')
                .addStringOption(option =>
                    option.setName('api_base')
                        .setDescription('Your local API URL (e.g., http://localhost:11434)')
                        .setRequired(true))
                .addStringOption(option =>
                    option.setName('api_key')
                        .setDescription('API key if required (leave empty if none)')
                        .setRequired(false))
                .addStringOption(option =>
                    option.setName('model_name')
                        .setDescription('Model name to use')
                        .setRequired(false)))
        .addSubcommand(subcommand =>
            subcommand
                .setName('prompt')
                .setDescription('Set a custom system prompt')
                .addStringOption(option =>
                    option.setName('prompt')
                        .setDescription('System prompt to prepend to all AI requests (leave empty to clear)')
                        .setRequired(false)))
        .addSubcommand(subcommand =>
            subcommand
                .setName('toggle')
                .setDescription('Enable or disable AI responses')
                .addBooleanOption(option =>
                    option.setName('enabled')
                        .setDescription('Enable/disable AI responses')
                        .setRequired(true)))
        .addSubcommand(subcommand =>
            subcommand
                .setName('channel')
                .setDescription('Add or remove a whitelisted channel')
                .addChannelOption(option =>
                    option.setName('channel')
                        .setDescription('Channel to whitelist')
                        .setRequired(true)
                        .addChannelTypes(ChannelType.GuildText))
                .addStringOption(option =>
                    option.setName('action')
                        .setDescription('Add or remove from whitelist')
                        .setRequired(true)
                        .addChoices(
                            { name: 'Add', value: 'add' },
                            { name: 'Remove', value: 'remove' }
                        )))
        .addSubcommand(subcommand =>
            subcommand
                .setName('listchannels')
                .setDescription('List all whitelisted channels'))
        .addSubcommand(subcommand =>
            subcommand
                .setName('chance')
                .setDescription('Set AI response chance percentage (0-100)')
                .addNumberOption(option =>
                    option.setName('percentage')
                        .setDescription('Chance percentage (0-100)')
                        .setRequired(true)
                        .setMinValue(0)
                        .setMaxValue(100)))
        .addSubcommand(subcommand =>
            subcommand
                .setName('test')
                .setDescription('Test your AI connection')
                .addStringOption(option =>
                    option.setName('message')
                        .setDescription('Test message to send to the AI')
                        .setRequired(true))),

    async execute(interaction) {
        const subcommand = interaction.options.getSubcommand();

        switch (subcommand) {
            case 'configure':
                await this.configureLocalAI(interaction);
                break;
            case 'prompt':
                await this.setSystemPrompt(interaction);
                break;
            case 'toggle':
                await this.toggleAI(interaction);
                break;
            case 'channel':
                await this.manageWhitelist(interaction);
                break;
            case 'listchannels':
                await this.listWhitelistedChannels(interaction);
                break;
            case 'chance':
                await this.setResponseChance(interaction);
                break;
            case 'test':
                await this.testAIConnection(interaction);
                break;
            default:
                await interaction.reply({
                    content: 'Unknown subcommand.',
                    flags: 64
                });
        }
    },

    async configureLocalAI(interaction) {
        const apiBase = interaction.options.getString('api_base');
        const apiKey = interaction.options.getString('api_key');
        const modelName = interaction.options.getString('model_name') || 'llama3';

        // Validate URL format
        if (!apiBase.startsWith('http://') && !apiBase.startsWith('https://')) {
            return await interaction.reply({
                content: 'Invalid URL format. Must start with http:// or https://',
                flags: 64
            });
        }

        // Save configuration to database
        await localAIModel.setConfig(
            interaction.guild.id,
            apiBase,
            apiKey,
            modelName
        );

        await interaction.reply({
            content: `Local AI configured!\nEndpoint: \`${apiBase}\`\nModel: \`${modelName}\`\nAPI Key: ${apiKey ? 'Configured' : 'None'}`,
            flags: 64
        });
    },

    async setSystemPrompt(interaction) {
        const prompt = interaction.options.getString('prompt');

        await localAIModel.setSystemPrompt(interaction.guild.id, prompt);

        if (prompt) {
            await interaction.reply({
                content: 'System prompt set! It will be prepended to all AI requests.',
                flags: 64
            });
        } else {
            await interaction.reply({
                content: 'System prompt cleared!',
                flags: 64
            });
        }
    },

    async toggleAI(interaction) {
        const enabled = interaction.options.getBoolean('enabled');

        await localAIModel.setEnabled(interaction.guild.id, enabled);

        await interaction.reply({
            content: `Local AI responses ${enabled ? 'enabled' : 'disabled'}`,
            flags: 64
        });
    },

    async manageWhitelist(interaction) {
        const channel = interaction.options.getChannel('channel');
        const action = interaction.options.getString('action');

        if (action === 'add') {
            await localAIModel.addWhitelistedChannel(interaction.guild.id, channel.id);
            await interaction.reply({
                content: `Added ${channel} to whitelist`,
                flags: 64
            });
        } else if (action === 'remove') {
            await localAIModel.removeWhitelistedChannel(interaction.guild.id, channel.id);
            await interaction.reply({
                content: `Removed ${channel} from whitelist`,
                flags: 64
            });
        }
    },

    async listWhitelistedChannels(interaction) {
        const channels = await localAIModel.getWhitelistedChannels(interaction.guild.id);
        
        if (!channels.length) {
            return await interaction.reply({
                content: 'No whitelisted channels',
                flags: 64
            });
        }

        const channelMentions = [];
        for (const channelId of channels) {
            const channel = interaction.guild.channels.cache.get(channelId);
            if (channel) {
                channelMentions.push(channel.toString());
            }
        }

        const embed = new EmbedBuilder()
            .setTitle('Whitelisted Channels')
            .setDescription(channelMentions.join('\n') || 'None')
            .setColor(0x7289DA);

        await interaction.reply({
            embeds: [embed],
            flags: 64
        });
    },

    async setResponseChance(interaction) {
        const chance = interaction.options.getNumber('percentage');

        await localAIModel.setResponseChance(interaction.guild.id, chance);

        await interaction.reply({
            content: `AI response chance set to ${chance}%`,
            flags: 64
        });
    },

    async testAIConnection(interaction) {
        await interaction.deferReply({ flags: 64 });

        const message = interaction.options.getString('message');
        const config = await localAIModel.getConfig(interaction.guild.id);

        if (!config) {
            return await interaction.editReply('AI is not configured. Use `/chatbot configure` first.');
        }

        try {
            // Get system prompt if configured
            const systemPrompt = await localAIModel.getSystemPrompt(interaction.guild.id);
            
            // Build messages array
            const messages = [];
            if (systemPrompt) {
                messages.push({ role: "system", content: systemPrompt });
            }
            messages.push({ role: "user", content: message });
            
            const headers = {};
            if (config.apiKey) {
                headers.Authorization = `Bearer ${config.apiKey}`;
            }

            // Call the AI API
            const response = await axios.post(
                `${config.apiBase}/v1/chat/completions`,
                {
                    model: config.modelName,
                    messages: messages,
                    temperature: config.temperature || 0.7,
                    max_tokens: config.maxTokens || 1000
                },
                {
                    headers: headers,
                    timeout: 30000
                }
            );

            if (response.status === 200 && response.data.choices && response.data.choices.length > 0) {
                const aiResponse = response.data.choices[0].message.content;
                
                // Truncate if needed
                const truncatedResponse = aiResponse.length > 1500 
                    ? aiResponse.substring(0, 1497) + '...' 
                    : aiResponse;
                
                const embed = new EmbedBuilder()
                    .setTitle('AI Test Response')
                    .setDescription(truncatedResponse)
                    .setColor(0x2B82CB)
                    .setFooter({ text: `Model: ${config.modelName}` })
                    .setTimestamp();
                
                await interaction.editReply({ embeds: [embed] });
            } else {
                await interaction.editReply('Received an unexpected response from the AI service.');
            }
        } catch (error) {
            console.error('AI API Error:', error);
            await interaction.editReply(`Error connecting to AI: ${error.message}`);
        }
    },

    /**
     * Get AI response for a message
     * @param {Client} client - Discord.js client
     * @param {Message} message - Discord message
     * @param {string} prompt - Prompt text
     * @returns {Promise<string|null>} - AI response or null
     */
    async getAIResponse(client, message, prompt) {
        try {
            const config = await localAIModel.getConfig(message.guild.id);
            if (!config || !config.enabled) {
                return null;
            }

            const headers = {};
            if (config.apiKey) {
                headers.Authorization = `Bearer ${config.apiKey}`;
            }

            // Get system prompt if configured
            const systemPrompt = await localAIModel.getSystemPrompt(message.guild.id);
            
            // Build messages array
            const messages = [];
            if (systemPrompt) {
                messages.push({ role: "system", content: systemPrompt });
            }
            messages.push({ role: "user", content: prompt });

            const response = await axios.post(
                `${config.apiBase}/v1/chat/completions`,
                {
                    model: config.modelName,
                    messages: messages,
                    temperature: config.temperature || 0.7,
                    max_tokens: config.maxTokens || 1000
                },
                {
                    headers: headers,
                    timeout: 30000
                }
            );

            if (response.status === 200 && response.data.choices && response.data.choices.length > 0) {
                return response.data.choices[0].message.content;
            }
            
            return null;
        } catch (error) {
            console.error('AI Request Failed:', error);
            return null;
        }
    }
};